File: app/src/main/java/ru/dmitrybochkov/rates/ui/rates/RatesActivity.java
package ru.dmitrybochkov.rates.ui.rates;

import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import butterknife.BindView;
import butterknife.ButterKnife;
import ru.dmitrybochkov.rates.R;
import ru.dmitrybochkov.rates.domain.Rate;
import ru.dmitrybochkov.rates.ui.BaseActivity;
import ru.dmitrybochkov.rates.view_models.RatesViewModel;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import java.util.List;

public class RatesActivity extends BaseActivity {

    @BindView(R.id.recycler_view)
    RecyclerView ratesList;
    @BindView(R.id.swipe_refresh_layout)
    SwipeRefreshLayout swipeRefreshLayout;
    @BindView(R.id.no_data)
    TextView noDataMessage;

    RatesViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_rates);
        ButterKnife.bind(this);
        viewModel = ViewModelProviders.of(this).get(RatesViewModel.class);
        initViews();
        bindData();
    }

    private void initViews() {
        ratesList.setLayoutManager(new LinearLayoutManager(this));
        ratesList.setAdapter(new RatesListAdapter());

        swipeRefreshLayout.setOnRefreshListener(() -> {
            viewModel.updateRates();
            swipeRefreshLayout.setRefreshing(false);
        });
    }

    private void bindData() {
        viewModel.getRates().observe(this, this::updateRatesList);
    }

    private void updateRatesList(List<Rate> rates) {
//        if (rates.isEmpty()) {
//            noDataMessage.setVisibility(View.VISIBLE);
//            swipeRefreshLayout.setVisibility(View.INVISIBLE);
//        } else {
//            noDataMessage.setVisibility(View.INVISIBLE);
//            swipeRefreshLayout.setVisibility(View.VISIBLE);
//        }
        ((RatesListAdapter) ratesList.getAdapter()).updateItems(rates);
    }
}

File: app/src/main/java/ru/dmitrybochkov/rates/ui/rates/RatesActivity.kt
package ru.dmitrybochkov.rates.ui.rates

import android.os.Bundle
import android.widget.TextView
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import butterknife.BindView
import butterknife.ButterKnife
import ru.dmitrybochkov.rates.R
import ru.dmitrybochkov.rates.domain.Rate
import ru.dmitrybochkov.rates.ui.BaseActivity
import ru.dmitrybochkov.rates.view_models.RatesViewModel

class RatesActivity : BaseActivity() {

    @BindView(R.id.recycler_view)
    internal lateinit var ratesList: RecyclerView
    @BindView(R.id.swipe_refresh_layout)
    internal lateinit var swipeRefreshLayout: SwipeRefreshLayout
    @BindView(R.id.no_data)
    internal lateinit var noDataMessage: TextView

    internal lateinit var viewModel: RatesViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_rates)
        ButterKnife.bind(this)
        viewModel = ViewModelProviders.of(this).get(RatesViewModel::class.java)
        initViews()
        bindData()
    }

    private fun initViews() {
        ratesList.layoutManager = LinearLayoutManager(this)
        ratesList.adapter = RatesListAdapter()

        swipeRefreshLayout.setOnRefreshListener {
            viewModel.updateRates()
            swipeRefreshLayout.isRefreshing = false
        }
    }

    private fun bindData() {
        viewModel.rates.observe(this, Observer<List<Rate>> { this.updateRatesList(it) })
    }

    private fun updateRatesList(rates: List<Rate>) {
        (ratesList.adapter as RatesListAdapter).updateItems(rates)
    }
}
